src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp

```diff
@@ -317,6 +317,17 @@ HRESULT STDMETHODCALLTYPE CorProfiler::ModuleLoadFinished(ModuleID module_id,
     return S_OK;
   }
 
+  // In IIS, the startup hook will be inserted into a method in System.Web (which is domain-neutral)
+  // but the Datadog.Trace.ClrProfiler.Managed.Loader assembly that the startup hook loads from a
+  // byte array will be loaded into a non-shared AppDomain.
+  // In this case, do not insert another startup hook into that non-shared AppDomain
+  if (module_info.assembly.name == "Datadog.Trace.ClrProfiler.Managed.Loader"_W) {
+    Info("ModuleLoadFinished: Datadog.Trace.ClrProfiler.Managed.Loader loaded into AppDomain ",
+          app_domain_id, " ", module_info.assembly.app_domain_name);
+    first_jit_compilation_app_domains.insert(app_domain_id);
+    return S_OK;
+  }
+
```

Comment:
Does this have any impact on .NET Core web applications?

Reply:
That's a good question. I don't know the answer right now but I will look into it.

Reply:
Based on this [documentation](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/iis/?view=aspnetcore-3.1), this should not affect the out-of-process hosting model since the application will be running in the `dotnet.exe` process. I'll look into the in-process hosting model to see what behavior, if any, is affected

Reply:
Sounds good. Thanks.

Reply:
Good news: It turns out that .NET Core web applications hosted in-process on IIS have the same startup code path as .NET Framework, so we solve the exact same issue in both places 👍

After the common code path, the .NET Core application is actually started and run inside a `clrhost` AppDomain, which triggers the startup hook to run inside that AppDomain. Currently, we're fortunate that we only run it once since it gets called by `Microsoft.Extensions.Hosting.GenericHostWebHostBuilderExtensions.ConfigureWebHost()` (on .NET Core 3.1) but this behavior was not changed. While we should make changes to be extremely explicit about where the startup hook is happening on aspnetcore applications, I think that could wait for a follow-up PR.

Reply:
It looks like the out-of-process model does the same thing except that the .NET Core application is started in a new `dotnet.exe` process instead of just a `clrhost` AppDomain